### PR TITLE
apply patch from pull/17731

### PR DESCRIPTION
**Pull request by detroitpro**
Apply patch from PR: https://github.com/ethereum/go-ethereum/pull/17731

- [ ] Validate tests
- [ ] Create test network (switches on a decided block #)

Help Wanted Test network
- [ ] bootnodes
- [ ] stats
- [ ] explorer

### CODE REVIEW

**Comment by mkrufky** [comment]
hardcore nitpick: "Apply patch from PR: ethereum/go-ethereum#17731" is not a useful description.   The idea of a git repository is that it and all of its metadata self contained.  One should be able to understand an applied patch (or pr) from looking at that patch (or pr) alone.  At least copy/paste the description from the other PR into this one, please.

File: params/config.go [hunk 1]
```diff
@@ -178,7 +178,7 @@ type ChainConfig struct {
 	BaneslayerBlock     *big.Int `json:"baneslayerBlock,omitempty"`     //second major akroma release
 	CopperLeafBlock     *big.Int `json:"copperLeafBlock,omitempty"`     //third major akroma release
 	DawnbreakBlock      *big.Int `json:"dawnbreakBlock,omitempty"`      //fourth major akroma release
-
+	ProgpowBlock        *big.Int `json:"progpowBlock,omitempty"`        // Progpow switch block (nil = not active, 0 = already activated)
```
Comment (mkrufky) [inline params/config.go]: I would have left the skip line and simply added the new `ProgpowBlock` above it.

File: consensus/ethash/progpow_test.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+}
+
+func TestCDag(t *testing.T) {
+	size := cacheSize(0)
+	cache := make([]uint32, size/4)
+	seed := seedHash(0)
+	generateCache(cache, 0, seed)
+	cDag := make([]uint32, progpowCacheWords)
+	generateCDag(cDag, cache, 0)
+	//fmt.Printf("Cdag: %d \n", cDag[:20])
```
Comment (mkrufky) [inline consensus/ethash/progpow_test.go]: was this useful?  should we enable it via a debug / verbosity setting?  or if it isn't useful, just erase it

File: consensus/ethash/progpow_test.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+}
+
+func TestProgpowChanges(t *testing.T) {
+	headerHash := common.HexToHash("ffeeddccbbaa9988776655443322110000112233445566778899aabbccddeeff")
+	nonce := uint64(0x123456789abcdef0)
+	blocknum := uint64(30000)
+	seed := seedHash(blocknum)
+	fmt.Printf("seedHash %x\n", seed)
+	//seed =  common.FromHex("ee304846ddd0a47b")
```
Comment (mkrufky) [inline consensus/ethash/progpow_test.go]: was this just debug during devel?  please remove if it isnt needed anymore

File: consensus/ethash/consensus_test.go [hunk 4]
```diff
@@ -84,3 +84,81 @@ func TestCalcDifficulty(t *testing.T) {
...
+//		}
+//		if _, err := bc.InsertChain(blocks[:i]); err != nil {
+//			t.Fatalf("invalid block %d: %v", n, err)
+//		}
+//	}
+//	if bc.CurrentBlock().Number().Cmp(big.NewInt(1054)) != 0 {
+//		t.Errorf("Expected to import 1054 blocks, got %v", bc.CurrentBlock().Number())
+//
+//	}
+//}
```
Comment (mkrufky) [inline consensus/ethash/consensus_test.go]: does it fail?  or fail to build?  we should fix the tests and make them pass

File: ProgPoW.diff [hunk 5]
```diff
@@ -0,0 +1,1314 @@
...
++++ b/tests/block_test_util.go
+@@ -116,7 +116,7 @@ func (t *BlockTest) Run() error {
+ 	if t.json.SealEngine == "NoProof" {
+ 		engine = ethash.NewFaker()
+ 	} else {
+-		engine = ethash.NewShared()
++		engine = ethash.NewShared(nil)
+ 	}
+ 	chain, err := core.NewBlockChain(db, &core.CacheConfig{TrieCleanLimit: 0}, config, engine, vm.Config{}, nil)
+ 	if err != nil {
```
Comment (mkrufky) [inline ProgPoW.diff]: why include this `diff` patch file?

File: consensus/ethash/ethash.go [hunk 6]
```diff
@@ -564,8 +574,15 @@ func NewFullFaker() *Ethash {
...
 // in the same process.
-func NewShared() *Ethash {
-	return &Ethash{shared: sharedEthash}
+func NewShared(progpowNumber *big.Int) *Ethash {
+	ethashMu.Lock()
+	sharedEngine, exist := sharedEngines[progpowNumber.Uint64()]
+	if !exist {
+		sharedEngine = New(Config{"", 3, 0, "", 1, 0, ModeNormal, progpowNumber}, nil, false)
+		sharedEngines[progpowNumber.Uint64()] = sharedEngine
+	}
```
Comment (mkrufky) [inline consensus/ethash/ethash.go]: This is questionable.  I didn't sort through the logic yet, but at first glance, it seems wrong for ANYTHING ProgPOW related to be referenced in the ethash code.

I see it's done in the upstream PR as well - I think it's wrong, and ugly.

**Comment by detroitpro** [comment]
@mkrufky - The node code seems pretty stable. There is another set of upstream changes that I need to merge in.

The miner code, is something I need to do more research into.

**Pull request closed without merging**